trezord: init at 1.2.0

canndrew opened the pull request:
###### Motivation for this change

I want to use my trezor with NixOS.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

canndrew committed:
trezord: init at 1.2.0

canndrew commented:
I'm very much a NixOS noob and barely know what I'm doing. So be sure to review this carefully.

LnL7 commented:
I noticed darwin related things in that patch, but it's marked as linux only.

joachifm commented on nixos/modules/misc/ids.nix:
```diff
@@ -284,6 +284,7 @@
       glance = 266;
       couchpotato = 267;
       gogs = 268;
+      trezord = 269;
```
What sort of data does trezord own?

canndrew replied:
It controls access to the trezor.

joachifm replied:
But the trezor is a hardware device, correct? Static uids are for ensuring consistent ownership across NixOS deployments.

joachifm replied:
Or if you need to refer to a user before the name/uid mapping is available.

canndrew replied:
@joachifm How do I use a non-static uid then? Do I just not give a uid/gid to `extraUsers`/`extraGroups`?

joachifm commented on nixos/modules/services/hardware/trezord.nix:
```diff
@@ -0,0 +1,71 @@
+        '';
+      };
+
+      user = mkOption {
```
What is the use case for running as a user other than the default?

canndrew replied:
None that I know of, but I thought I'd make it flexible.

joachifm replied:
To my mind, adding options "just because" is not real flexibility, but rather bloats the configuration state space to no benefit. Absent a plausible use case for running a daemon under a different user, please re-consider whether it needs to be possible to do so. This holds doubly true for daemons that can work with dynamic uids, as down the line we can rely on systemd to do that for us, obviating the need to declare daemon users up-front for those services.

canndrew replied:
Alright, I've removed the ability to run that daemon as anything other than the `trezord` user.

canndrew commented:
> I noticed darwin related things in that patch, but it's marked as linux only.

The line is `if (`**`NOT`**` ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")`, so it effects non-darwin systems.

canndrew commented on pkgs/servers/trezord/default.nix:
```diff
@@ -0,0 +1,51 @@
+    description = "TREZOR Bridge daemon for TREZOR bitcoin hardware wallet";
+    homepage = https://mytrezor.com;
+    license = licenses.gpl3;
+    maintainers = [];
```
I'll add both of us.

canndrew committed:
trezord: minor fixes

* Use fetchFromGitHub
* Add maintainers

canndrew commented on pkgs/servers/trezord/default.nix:
```diff
@@ -0,0 +1,51 @@
+stdenv.mkDerivation rec {
+  name = "trezord-${version}";
+
+  src = fetchgit {
```
Done.

canndrew replied:
Okay, I tried using `fetchFromGitHub` but for some reason it was grabbing a different version with a different hash which for some reason didn't build. So I've gone back to `fetchgit`.

joachifm replied:
`fetchgit` should only be necessary if you need submodules. Can you elaborate on the build error?

canndrew replied:
Ah yes, the repo does have submodules in it. That's what's causing it.

canndrew committed:
trezord: more fixes

* Remove static uid/gid
* Use fetchgit rather than fetchFromGitHub.
  It seems to be grapping a different version with fetchFromGGitHub for some
  reason.
* Always create a user/group named trezord

canndrew committed:
trezord: Add canndrew to maintainers list

joachifm commented on nixos/modules/services/hardware/trezord.nix:
```diff
@@ -0,0 +1,57 @@
+      };
+    };
+
+    users.extraUsers = singleton {
```
Can also be written
```
users.users.trezord = {
  group = "trezord";
  description = /* ... */;
};
```
When name is `null`, it is taken from the attrset key.

joachifm commented on nixos/modules/services/hardware/trezord.nix:
```diff
@@ -0,0 +1,57 @@
+      description = "Trezor bridge daemon user";
+    };
+
+    users.extraGroups = singleton {
```
Can also be written
```nix
let # for syntax hl
users.groups.trezord = {};
```

joachifm commented on nixos/modules/services/hardware/trezord.nix:
```diff
@@ -0,0 +1,57 @@
+      description = "TREZOR Bridge";
+      after = [ "systemd-udev-settle.service" "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      path = [ pkgs.trezord ];
```
Note that it is not necessary to add stuff to `path`, if it's just used in the `ExecStart`.

canndrew commented:
Is this waiting for me to make changes or is okay to merge?

canndrew committed:
trezord: update with joachifm's suggestions

canndrew commented:
I've updated with @joachifm's suggestions. Hopefully this can be merged now.
